WIP: Stacklight integration

cznewt opened the pull request.

Commit:
Stacklight integration

simonpasquier commented on heka/files/toml/decoder/sandbox.toml:
```diff
@@ -0,0 +1,63 @@
+output_limit = "{{ decoder.output_limit }}"
+{%- endif %}
+
+[{{ decoder_name }}_decoder.config]
```
Hmm, I would do it differently. A Lua sandbox plugin (decoder, filter, encoder) is able to pass anything in the [<plugin>.config] section. From the Heka documentation:

> config (object): A map of configuration variables available to the sandbox via read_config. The map consists of a string key with: string, bool, int64, or float64 values.

This is how we've done in the Puppet module => https://github.com/openstack/fuel-plugin-lma-collector/blob/master/deployment_scripts/puppet/modules/heka/templates/decoder/sandbox.toml.erb

I'd imagine that it is possible to do the same in Jinja2 style with Salt passing a dict in the "decoder.config" key. The same applies for the filter and encoder templates.

cznewt replied:
Implemented as you say, sample metadata changed accordingly

simonpasquier commented on heka/files/toml/global.toml:
```diff
@@ -0,0 +1,12 @@
+max_message_size = 262144
+max_process_inject = 1
+max_timer_inject = 10
+poolsize = 100
```
These parameters should be templatized eventually but it's not required yet.

cznewt replied:
yes, anyone feel free to parametrise these :smile:

simonpasquier commented on heka/meta/heka.yml:
```diff
@@ -0,0 +1,80 @@
+      module_dir: /usr/share/lma_collector_modules;/usr/share/heka/lua_modules
+      include_full_notification: 'false'
+  input:
+    cinder_log_stream:
```
Should go to the Cinder formula?

cznewt replied:
yes

simonpasquier commented on heka/meta/heka.yml:
```diff
@@ -0,0 +1,80 @@
+      decoder: "openstack_decoder"
+      splitter: "openstack_splitter"
+  filter:
+    linux_hdd_errors:
```
Should go to the linux formula?

cznewt replied:
yes, to debug purpose I put it here, it will be moved quickly.

elemoine commented on heka/_service.sls:
```diff
@@ -0,0 +1,209 @@
+{%- macro service_config(service_name) %}
+
+{%- set server = salt['pillar.get']('heka:'+service_name) %}
```
Super minor but I'd add spaces around the "+" sign, for a better readability.

elemoine commented on heka/_service.sls:
```diff
@@ -0,0 +1,209 @@
+  - name: /etc/init/{{ service_name }}.conf
+  - user: root
+  - mode: 644
+  - template: jinja
```
In this declaration exactly the same as the previous one? If that's the case it can probably be moved outside the if/else block.

elemoine commented on heka/_service.sls:
```diff
@@ -0,0 +1,209 @@
+
+{%- endfor %}
+
+{%- for service_name, service in pillar.items() %}
```
You already have a more global `service_name` variable. I think it would be better to use two different variable names.

elemoine commented on heka/_service.sls:
```diff
@@ -0,0 +1,209 @@
+  - require_in:
+    - file: heka_{{ service_name }}_conf_dir_clean
+  - watch_in:
+    - service: heka_{{ service_name }}_service
```
You will need to change that. The services should somehow be able to specify different Heka config files based on the Heka role ("aggregator" versus "log_collector" etc.).

elemoine commented on heka/meta/heka.yml:
```diff
@@ -0,0 +1,85 @@
+log_collector:
```
You created an `heka/meta/heka.yml` while `heka` is still disabled in the `support` reclass. Do you mean to enable the heka support in `metadata/service/support.yml`?

SwannCroiset commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+{%- macro load_grains_file(grains_fragment_file) %}{% include grains_fragment_file %}{% endmacro %}
```
Since not all services have meta/heka.yml until you add it to all formulas, this macro will fail.
Thanks to jinja2 for ignoring missing:

{%- macro load_grains_file() %}{% include grains_fragment_file ignore missing %}{% endmacro %}

fpytloun commented on heka/_common.sls:
```diff
@@ -0,0 +1,29 @@
+  pkg.latest:
+  - names: {{ server.pkgs }}
+
+/usr/share/lma_collector:
```
Such custom things belong to `/usr/local` prefix, so `/usr/local/share/lma_collector`

fpytloun commented on heka/_common.sls:
```diff
@@ -0,0 +1,29 @@
+  user.present:
+  - name: heka
+  - system: true
+  - shell: /bin/nologin
```
It's `/usr/sbin/nologin` or `/bin/false` which seems to be more common:

```
$ grep -c nologin /etc/passwd                                                                                  (master*)  
17
$ grep -c false /etc/passwd                                                                                    (master*)  
28
```

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+  file.managed:
+  - name: /var/log/{{ service_name }}.log
+  - user: heka
+  - mode: 750
```
Execution bit on log file?

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+heka_{{ service_name }}_conf_dir:
+  file.directory:
+  - name: /etc/{{ service_name }}
+  - user: root
```
Shouldn't it be `root:heka` to be readable by heka group?

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+heka_{{ service_name }}_service_file:
+  file.managed:
+  - name: /etc/init/{{ service_name }}.conf
```
This isn't location for systemd but upstart (according to conditional above you are doing this on systemd-enabled system).
Also source is missing here.

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+heka_{{ service_name }}_service_file:
+  file.managed:
+  - name: /etc/init/{{ service_name }}.conf
```
Again missing source and yes this is upstart so there should be conditional to setup only on upstart-enabled systems.

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+heka_{{ service_name }}_service_wrapper:
+  file.managed:
+  - name: /usr/local/bin/{{ service_name }}.log
```
So log file or binary (.log extention)? :-)

fpytloun replied:
Also no source again

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+{%- endif %}
+
+
+{# Setup basic structure for all roles so updates can apply #}
```
Move this to map.jinja? Don't know if it will be more readable.

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+{%- set grains_fragment_file = service_name+'/meta/heka.yml' %}
+{%- set grains_yaml = load_grains_file(grains_fragment_file)|load_yaml %}
+{%- set _dummy = service_grains.heka.update(grains_yaml) %}
```
Dummy is not needed, use

```
{%- do service_grains.heka.update(grains_yaml) %}
```

SwannCroiset replied:
good to know, this pattern is used everywhere in other formulas btw (collectd, ..)

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+{%- for node_name, node_grains in salt['mine.get']('*', 'grains.items').iteritems() %}
+{%- if node_grains.heka is defined %}
+
+{%- set _dummy = service_grains.heka.update(node_grains.heka) %}
```
Again, dummy not needed

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+{%- for service_action_name, service_action in service_grain.items() %}
+{%- if salt['pillar.get']('heka:'+service_grain_name).get(service_action_name, False) is mapping %}
+{%- set _dummy = service_grains.heka.[service_grain_name].[service_action_name].update(salt['pillar.get']('heka:'+service_grain_name+':'+service_action_name)) %}
```
same here

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+{%- for service_grain_name, service_grain in service_grains.items() %}
+{% if salt['pillar.get']('heka:'+service_grain_name) %}
+
+{%- for service_action_name, service_action in service_grain.items() %}
```
Maybe better use .iteritems everywhere

fpytloun commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+{%- for decoder_name, decoder in service_grains.heka.[service_name].decoder.iteritems() %}
+
+/etc/{{ service_name }}/10-decoder-{{ decoder_name }}-{{ decoder.engine }}.toml:
```
You could maybe use subdirectory for decoders to avoid having directory with tons of files.

fpytloun commented on heka/files/heka.service:
```diff
@@ -1,18 +1,44 @@
 [Install]
 WantedBy=multi-user.target
 
+{%- else %}
```
I would prefer spliting init scripts per type into separate files than doing more jinja conditionals

fpytloun replied:
Also this is upstart, you should be specific in that as you don't want to deploy this on standard sysv init systems

SwannCroiset commented on heka/_service.sls:
```diff
@@ -0,0 +1,269 @@
+
+{%- set grains_fragment_file = service_name+'/meta/heka.yml' %}
+{%- set grains_yaml = load_grains_file(grains_fragment_file)|load_yaml %}
+{%- set _dummy = service_grains.heka.update(grains_yaml) %}
```
L94 +{%- set _dummy = service_grains.heka.update(grains_yaml) %}

here the last service wins: you are erasing "service_grains" entries with the last one

Commit:
Round 2

elemoine commented on heka/_service.sls:
```diff
@@ -0,0 +1,302 @@
+
+{%- set grains_fragment_file = service_name+'/meta/heka.yml' %}
+{%- set grains_yaml = load_grains_file(grains_fragment_file)|load_yaml %}
+{%- do service_grains.heka.update(grains_yaml) %}
```
I get an error here when applying the `heka` state:

```
CRITICAL] Rendering SLS 'base:heka.metric_collector' failed: Jinja error: 'NoneType' object is not iterable
/var/cache/salt/minion/files/base/heka/_service.sls(115):
---
[...]

{%- macro load_grains_file(grains_fragment_file) %}{% include grains_fragment_file %}{% endmacro %}

{%- set grains_fragment_file = service_name+'/meta/heka.yml' %}
{%- set grains_yaml = load_grains_file(grains_fragment_file)|load_yaml %}
{%- do service_grains.heka.update(grains_yaml) %}    <======================
```

elemoine commented on heka/_service.sls:
```diff
@@ -52,61 +50,68 @@ heka_{{ service_name }}_service_wrapper:
 {%- for service_name, service in pillar.iteritems() %}
 {%- if service.get('_support', {}).get('heka', {}).get('enabled', False) %}
 
+{%- macro load_grains_file(grains_fragment_file) %}{% include grains_fragment_file %}{% endmacro %}
```
I think you need to use `ignore missing` here, otherwise it will fail with for example the `salt` service which does support `heka` but does not have an `heka.yml` meta file.

http://jinja.pocoo.org/docs/dev/templates/#include

But note that this is not sufficient. After I add `ignore missing` I still get an error:

```
----------
          ID: heka_metric_collector_service_file
    Function: file.managed
        Name: /etc/systemd/system/metric_collector.service
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1733, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1652, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/states/file.py", line 1644, in managed
                  **kwargs
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 3953, in check_managed_changes
                  **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 3623, in get_managed
                  **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 178, in render_tmpl
                  output = render_str(tmplstr, context, tmplpath)
                File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 386, in render_jinja_tmpl
                  buf=tmplstr)
              SaltRenderError: Jinja variable 'service_name' is undefined
     Started: 10:13:51.733850
    Duration: 13.324 ms
     Changes:   
----------
          ID: heka_metric_collector_service
    Function: service.running
        Name: metric_collector
      Result: False
     Comment: The named service metric_collector is not available
     Started: 10:13:51.749779
    Duration: 14.005 ms
     Changes:
```

elemoine replied:
See https://github.com/tcpcloud/salt-formula-heka/pull/3 and https://github.com/tcpcloud/salt-formula-heka/pull/4.

Commit:
Variable service_name is missing for systemd file

Commit:
Use "ignore missing" with the j2 include statement

Commit:
preserve_data and ticker_interval are not strings

preserve_data is a boolean, and ticker_interval is a number, so their values
shouldn't have quotes.

cznewt committed:
Merge pull request #4 from elemoine/stacklight-ignore-missing

Use "ignore missing" with the j2 include statement

cznewt committed:
Merge pull request #3 from elemoine/stacklight-systemd

Variable service_name is missing for systemd file

cznewt committed:
Merge pull request #5 from elemoine/stacklight-sandbox-toml

preserve_data and ticker_interval are not strings

Commit:
Added cache dir

Commit:
Use the proper module directory

The stacklight module dir is /usr/share/lma_collector/common, not
/usr/share/lma_collector_modules. This fixes it.

Commit:
Add the extra_fields.lua module

This commit adds the extra_fields Lua module. The extra fields table defined in
this module is empty right now. Eventually, this file will be a Jinja2 template
and the content of the extra fields table will be generated based on the user
configuration.

cznewt committed:
Merge pull request #8 from elemoine/stacklight-module-dir-2

Use the proper module directory

cznewt committed:
Merge pull request #9 from elemoine/stacklight-extra-fields

Add the extra_fields.lua module

Commit:
Regex encoder fix

Commit:
Merge branch 'stacklight' of github.com:tcpcloud/salt-formula-heka into stacklight
